includes/class-events-store.php

```diff
@@ -0,0 +1,592 @@
+	 */
+	public function update_option( $new_value, $old_value ) {
+		// Find changes to record
+		$new_events     = $this->find_cron_array_differences( $new_value, $old_value );
+		$deleted_events = $this->find_cron_array_differences( $old_value, $new_value );
+
+		// Add/update new events
+		foreach ( $new_events as $new_event ) {
+			$job_id = $this->job_exists( $new_event['timestamp'], $new_event['action'], $new_event['instance'], true );
+
+			if ( 0 === $job_id ) {
+				$job_id = null;
+			}
+
+			$this->create_or_update_job( $new_event['timestamp'], $new_event['action'], $new_event['args'], $job_id, false );
+		}
+
+		// Mark deleted entries for removal
+		foreach ( $deleted_events as $deleted_event ) {
+			$this->mark_job_completed( $deleted_event['timestamp'], $deleted_event['action'], $deleted_event['instance'], false );
+		}
+
+		$this->flush_internal_caches();
+
+		return $old_value;
+	}
+
+	/**
+	 * When an entry exists, don't try to create it again
+	 */
+	public function block_creation_if_job_exists( $job ) {
+		$instance = md5( maybe_serialize( $job->args ) );
+		if ( $this->job_exists( $job->timestamp, $job->hook, $instance ) ) {
+			return false;
+		}
+
+		return $job;
+	}
+
+	/**
+	 * PLUGIN UTILITY METHODS
+	 */
+
+	/**
+	 * Retrieve jobs given a set of parameters
+	 *
+	 * @param array $args
+	 * @return array|false
+	 */
+	public function get_jobs( $args ) {
+		global $wpdb;
+
+		if ( ! isset( $args['quantity'] ) || ! is_numeric( $args['quantity'] ) ) {
+			$args['quantity'] = 100;
+		}
+
+		if ( isset( $args['page'] ) ) {
+			$page  = max( 0, $args['page'] - 1 );
+			$offset = $page * $args['quantity'];
+		} else {
+			$offset = 0;
+		}
+
+		// Do not sort, otherwise index isn't used
+		$jobs = $wpdb->get_results( $wpdb->prepare( "SELECT * FROM {$this->get_table_name()} WHERE status = %s LIMIT %d,%d;", $args['status'], $offset, $args['quantity'] ), 'OBJECT' );
+
+		if ( is_array( $jobs ) ) {
+			$jobs = array_map( array( $this, 'format_job' ), $jobs );
+		} else {
+			$jobs = false;
+		}
+
+		return $jobs;
+	}
+
+	/**
+	 * Retrieve a single event by ID, or by a combination of its timestamp, instance identifier, and either action or the action's hashed representation
+	 *
+	 * @param  array $attrs Array of event attributes to query by
+	 * @return object|false
+	 */
+	public function get_job( $attrs ) {
+		global $wpdb;
+
+		// Validate basic inputs
+		if ( ! is_array( $attrs ) || empty( $attrs ) ) {
+			return false;
+		}
+
+		// Validate requested status
+		$allowed_status = $this->get_allowed_statuses();
+		$allowed_status[] = 'any';
+
+		if ( ! isset( $attrs['status'] ) || ! in_array( $attrs['status'], $allowed_status, true ) ) {
+			$attrs['status'] = self::STATUS_PENDING;
+		}
+
+		// Validate attributes provided to query for a post
+		$query              = null;
+		$query_replacements = array();
+
+		if ( isset( $attrs['ID'] ) ) {
+			$query                = "SELECT * FROM {$this->get_table_name()} WHERE ID = %d";
```

Comment:
My intent was to make this method a bit like `WP_Query` in its flexibility. The use case for looking a job up by ID is limited to the plugin's CLI commands, so I can move this bit somewhere specific to that context.